docs/code/SchemaComprehension.md

```diff
@@ -0,0 +1,210 @@
+
+```(csharp)
+var cursorable = dv.AsCursorable<IrisVectorData>(env);
+// You can create as many simultaneous cursors as you like, they are independent.
+using (var cursor = cursorable.GetCursor())
+{
+    // We are now in charge of creating the row object.
+    var myRow = new IrisVectorData();
+    while (cursor.MoveNext())
+    {
+        if (cursor.Position % 100 == 99)
+        {
+            // Populate the values of the row object.
+            cursor.FillValues(myRow);
+            // Do something to the row.
+        }
+    }
+}
+``` 
+Please note that **cursors are not thread-safe**: they have mutable state inside, and they are meant to be used by one thread. If you want to read the data in parallel, use multiple cursors.
+
+## PredictionEngine and PredictorModel
+
+ML.NET's `PredictionEngine` is attempting to turn a sequence of data transforms (maybe capped by a predictor, but not necessarily) into a 'black box' that takes strongly typed inputs and returns strongly typed outputs. The name is a little misleading: the `PredictionEngine` object doesn't require a predictor to be present in the pipeline, it can be just a sequence of transforms like in the below example:
+
+```(csharp)
+var engine = env.CreatePredictionEngine<IrisData, IrisVectorData>(dv);
+var output = engine.Predict(new IrisData { Label = 1, PetalLength = 1, SepalLength = 1, PetalWidth = 1, SepalWidth = 1 });
+```
+It is important to note that the `PredictionEngine` actually *validates* that the 'pipeline' conforms to the input and output schema requirements when it is created.
+
+The same can be said about the `PredictorModel<InputType, OutputType>`. This is a somewhat more restricted version of `PredictionEngine` that is created by `LearningPipeline.Train`.
+
+Please note that **`PredictionEngine` and `PredictorModel` are not thread-safe**: they hold an internal cursor object, and therefore cannot be used in a re-entrant fashion.
+If you ever see the error message that says: `An attempt was made to keep iterating after the pipe has been reset`, it most likely means that ML.NET has detected a race condition on the `PredictionEngine`.
+
+## Type system mapping
+
+`IDataView` [type system](IDataViewTypeSystem.md) differs slightly from the C# type system, so a 1-1 mapping between column types and C# types is not always feasible. 
+Below are the most notable examples of the differences:
+
+* `IDataView` vector columns may have a fixed (and known) size, C# arrays can not. You can use `[VectorType(N)]` attribute to an array field to specify that the column is a vector of fixed size N. This is often necessary: most ML components don't work with variable-size vectors, they require fixed-size ones.
+* `IDataView`'s **key types** don't have an underlying C# type either. To declare a key-type column, you need to make your field an `uint`, and decorate it with `[KeyType(Min=A, Count=B)]` to denote that the field is a key with the specified range of values.
```

Comment:
This makes it sound like `Min` and `Count` are required values, but they do not appear to be.  (And I'm assuming there are plenty of scenarios where a user doesn't know up front what are all the possible values.